torch/nn/parallel/distributed.py

```diff
@@ -688,19 +688,17 @@ def __setstate__(self, state):
     def _build_params_for_reducer(self):
         # Build tuple of (module, parameter) for all parameters that require grads.
         modules_and_parameters = [
```

Comment:
We might need to update the subsequent usage on `modules_and_parameters` as well. Below are a few, there might be more.

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L709-L714

change to:

```python
modules_and_parameters = [
    (m, p) for m, p in modules_and_parameters 
    if p not in memo and not memo.add(p)
]
```

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L717-L720

change to:

```python
parameters = list(parameter for _, parameter in modules_and_parameters)
```

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L732-L735

change to:

```python
expect_sparse_gradient = list(produces_sparse_gradient(module) for module, _ in modules_and_parameters)
```

Reply:
Made all of these changes, thanks @mrshenli!